concordium-consensus/src/Concordium/KonsensusV1/Consensus/CatchUp.hs

```diff
@@ -13,3 +54,490 @@ module Concordium.KonsensusV1.Consensus.CatchUp where
+    * N's current epoch is at least P's current epoch.
+    * N has all timeout signatures on round(N) that P has, and considers them valid if they are.
+    * N has all quorum signatures on round(N) that P has, and considers them valid if they are.
+    * If P has only one block in round(N), then N also has that block.
+
+If P is honest and considers a timeout signature on round(N) valid, then
+-}
+
+-- |'CatchUpPartialResponse' represents a stream of blocks to send as a response to a catch-up
+-- request. The catch up response message should include the catch-up terminal data if all required
+-- blocks are sent. As the stream of blocks may be cut off early (e.g. to send max 100 blocks),
+-- if the last result is 'CatchUpPartialResponseBlock' then 'cuprFinish' should be called to
+-- determine if the terminal data should be included.
+data CatchUpPartialResponse m
+    = -- |The next block in the stream.
+      CatchUpPartialResponseBlock
+        { -- |Next block.
+          cuprNextBlock :: SignedBlock,
+          -- |Continuation for getting any further blocks.
+          cuprContinue :: m (CatchUpPartialResponse m),
+          -- |Continuation that gets the terminal data in the case where there are no further
+          -- blocks.
+          cuprFinish :: m (Option CatchUpTerminalData)
+        }
+    | -- |There are no further blocks, just the terminal data.
+      CatchUpPartialResponseDone
+        { -- |Terminal data.
+          cuprFinished :: CatchUpTerminalData
+        }
+
+-- |Construct a 'TimeoutSet' summary from 'TimeoutMessages'.
+makeTimeoutSet :: TimeoutMessages -> TimeoutSet
+makeTimeoutSet TimeoutMessages{..} =
+    TimeoutSet
+        { tsFirstEpoch = tmFirstEpoch,
+          tsFirstEpochTimeouts = finalizerSet (Map.keys tmFirstEpochTimeouts),
+          tsSecondEpochTimeouts = finalizerSet (Map.keys tmSecondEpochTimeouts)
+        }
+
+-- |Generate a catch up status for the current state of the consensus.
+makeCatchUpStatus :: SkovData pv -> CatchUpStatus
+makeCatchUpStatus sd = CatchUpStatus{..}
+  where
+    lfBlock = sd ^. lastFinalized
+    cusLastFinalizedBlock = getHash lfBlock
+    cusLastFinalizedRound = blockRound lfBlock
+    cusCurrentRound = sd ^. roundStatus . rsCurrentRound
+    cusCurrentRoundQuorum =
+        view _3
+            <$> sd ^. currentQuorumMessages . smBlockToWeightsAndSignatures
+    cusCurrentRoundTimeouts = makeTimeoutSet <$> sd ^. currentTimeoutMessages
+    getLeavesBranches _ curLeaves curBranches Seq.Empty = (curLeaves, curBranches)
+    getLeavesBranches parentBranches curLeaves curBranches (rest Seq.:|> cur) =
+        getLeavesBranches newParentBranches newLeaves newBranches rest
+      where
+        (newLeaves, newBranches, newParentBranches) = foldl' up (curLeaves, curBranches, Set.empty) cur
+        up (l, b, p) blk
+            | Set.member hsh parentBranches = (l, hsh : b, p')
+            | otherwise = (hsh : l, b, p')
```

Comment:
It would help readability if you provided a bit of a description of this. A type and description of arguments would suffice I think.